### DISCUSSION TITLE
How to enable custom #define for ADE7880 driver

### DISCUSSION THREAD

**Discussion by AleXSR700** [arendst/Tasmota#19391] [Support request]
Hello everyone,

I would really like to not just ask for help but be part of the solution.
But well, I have my IT strengths and weaknesses.

I would like to adjust xnrg_23_ade7880.ino so the user can `#define` the different calibration data during compilation rather than using (and wasting) rules.

When I try to `#undef` and then `#define` via `user_config_override.h`
```
#ifdef FIRMWARE_SHELLY3EM
    #undef USE_ADE7880                  // Add support for ADE7880 for Shelly 3EM
    #define USE_ADE7880                 // Add support for ADE7880 for Shelly 3EM
    #undef ADE7880_AIGAIN_INIT          // rms, current_a
```
I got
```
Compiling .pio/build/tasmota-shelly3em/src/tasmota.ino.cpp.o
/tasmota/tasmota/tasmota_xnrg_energy/xnrg_23_ade7880.ino:48:0: warning: "ADE7880_AIGAIN_INIT" redefined [enabled by default]
 #define ADE7880_AIGAIN_INIT     3166385      // rms, current_a
 ```
 So sfromis told me that the driver does not support this.
 So I changed the driver to
 ```
 #define ADE7880_FREQ_INIT       0            // Connected to networks with fundamental frequencies between 55 Hz and 66 Hz (1). Default 45 Hz to 55 Hz (0).
#ifndef ADE7880_AIGAIN_INIT
  #define ADE7880_AIGAIN_INIT     3166385      // rms, current_a
#endif
```
(of course I always change all parameters but I omit them here)
Having done that, I run into the issue that the driver later needs this definition
```
/tasmota/tasmota/tasmota_xnrg_energy/xnrg_23_ade7880.ino: In function 'void Ade7880Show(bool)':
/tasmota/tasmota/tasmota_xnrg_energy/xnrg_23_ade7880.ino:777:94: error: 'EnergyFormat' was not declared in this scope
       EnergyFormat(value_chr, &Ade7880.neutral_current, Settings->flag2.current_resolution, 1));
```
Could someone help me or tell me what to do to make it work?
I would be happy to do any leg work/busy work that is needed to make this a default for everyone.

Also happy to grant rights to anyone willing to help, if that would make it easier for some
https://github.com/AleXSR700/Tasmota-dev/blob/development/tasmota/tasmota_xnrg_energy/xnrg_23_ade7880.ino

Thank you :)

**Comment by arendst** [comment]
I saw your discord discussion and hoped you never got here....

Anyway. I suggest you first try to make your FIRMWARE_SHELLY3EM functional without changing the ADE7880 driver.
Once that's done you should be able to change ADE7880_AIGAIN_INIT like you did.

I guess your current issue with the EnergyFormat error is that in addition to your enabled ADE7880 driver you also need the energy driver (define USE_ENERGY_SENSOR) which provides the EnergyFormat function (and the core of any energy presentation). And ofcourse the ADE7990 needs I2C enable (define USE_I2C)

In addition I made your suggested changes to the ADE7880 driver in the latest dev release just now.

**Comment by arendst** [comment]
Oh, looking at your file I noticed you are not using the latest released driver source. It has changed some months ago.

I suggest you restart your effort using the latest dev branch which has the correct EnergyFmt function.

**Comment by AleXSR700** [comment]
> I saw your discord discussion and hoped you never got here....

No such luck, sorry.

But I hope it will be quick and painless.

I added I2C and energy_sensor and pulled the git again.

I have no idea why my fork was outdated. I literally forked it yesterday.

Anyway, build went through and I am now testing the "normal" build on a spare device before testing the Shelly 3EM build.
